pkg/l2announcer/l2announcer.go

```diff
@@ -0,0 +1,1073 @@
+	devicesChan   chan []string
+
+	// selectedPolicies matching the current node.
+	selectedPolicies map[resource.Key]*selectedPolicy
+	// Services which are selected by one or more policies for which we thus want to participate in leader election.
+	// Indexed by service key.
+	selectedServices map[resource.Key]*selectedService
+	// A list of devices which can be matched by the policies
+	devices []string
+}
+
+func NewL2Announcer(params l2AnnouncerParams) *L2Announcer {
+	// These values were picked because it seemed right, change if necessary
+	const leaderElectionBufferSize = 16
+	const devicesBufferSize = 16
+	announcer := &L2Announcer{
+		Params: params,
+		jobgroup: params.JobRegistry.NewGroup(
+			job.WithLogger(params.Logger),
+			job.WithPprofLabels(pprof.Labels("cell", "l2-announcer")),
+		),
+		selectedServices: make(map[resource.Key]*selectedService),
+		selectedPolicies: make(map[resource.Key]*selectedPolicy),
+		leaderChannel:    make(chan leaderElectionEvent, leaderElectionBufferSize),
+		devicesChan:      make(chan []string, devicesBufferSize),
+	}
+
+	params.Lifecycle.Append(announcer.jobgroup)
+
+	if !params.DaemonConfig.EnableL2Announcements {
+		// If the L2 announcement feature is disabled, garbage collect any leases from previous runs when the feature
+		// might have been active. Just once, not on a timer.
+		announcer.jobgroup.Add(job.OneShot("l2-announcer lease-gc", announcer.leaseGC))
+		return announcer
+	}
+
+	announcer.jobgroup.Add(job.OneShot("l2-announcer run", announcer.run))
+	announcer.jobgroup.Add(job.Timer("l2-announcer lease-gc", announcer.leaseGC, time.Minute))
+
+	return announcer
+}
+
+// DevicesChanged can be invoked by an external component responsible for discovering all available network devices to
+// inform this component of all the devices we can use for L2 announcements.
+func (l2a *L2Announcer) DevicesChanged(devices []string) {
+	select {
+	// Send the change to the central routine to avoid race conditions
+	case l2a.devicesChan <- devices:
+	default:
+	}
+}
+
+func (l2a *L2Announcer) run(ctx context.Context) error {
+	var err error
+	l2a.svcStore, err = l2a.Params.Services.Store(ctx)
+	if err != nil {
+		return fmt.Errorf("get service store: %w", err)
+	}
+
+	l2a.policyStore, err = l2a.Params.L2AnnouncementPolicy.Store(ctx)
+	if err != nil {
+		return fmt.Errorf("get policy store: %w", err)
+	}
+
+	svcChan := l2a.Params.Services.Events(ctx)
+	policyChan := l2a.Params.L2AnnouncementPolicy.Events(ctx)
+	localNodeChan := l2a.Params.LocalNodeResource.Events(ctx)
+
+	// We have to first have a local node before we can start processing other events.
+	for {
+		select {
+		case event := <-localNodeChan:
+			if err := l2a.processLocalNodeEvent(ctx, event); err != nil {
+				l2a.Params.Logger.WithError(err).Error("Error processing local node event")
+			}
+		}
+
+		if l2a.localNode != nil {
+			break
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+		case event := <-svcChan:
```

Comment:
We don't process the channel closing. I think it's fine as long as we bail out from `ctx.Done`. Since we don't currently, I expect this turns into a busy-loop when one of the channels close.